icu4j/main/classes/core/src/com/ibm/icu/number/FormattedNumberRange.java

```diff
@@ -0,0 +1,209 @@
+// © 2018 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+package com.ibm.icu.number;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.text.AttributedCharacterIterator;
+import java.text.FieldPosition;
+import java.util.Arrays;
+
+import com.ibm.icu.impl.number.DecimalQuantity;
+import com.ibm.icu.impl.number.NumberStringBuilder;
+import com.ibm.icu.number.NumberRangeFormatter.RangeIdentityResult;
+import com.ibm.icu.util.ICUUncheckedIOException;
+
+/**
+ * The result of a number range formatting operation. This class allows the result to be exported in several data types,
+ * including a String, an AttributedCharacterIterator, and a BigDecimal.
+ *
+ * @author sffc
+ * @draft ICU 63
+ * @provisional This API might change or be removed in a future release.
+ * @see NumberRangeFormatter
+ */
+public class FormattedNumberRange {
+    final NumberStringBuilder string;
+    final DecimalQuantity quantity1;
+    final DecimalQuantity quantity2;
+    final RangeIdentityResult identityResult;
+
+    FormattedNumberRange(NumberStringBuilder string, DecimalQuantity quantity1, DecimalQuantity quantity2,
+            RangeIdentityResult identityResult) {
+        this.string = string;
+        this.quantity1 = quantity1;
+        this.quantity2 = quantity2;
+        this.identityResult = identityResult;
+    }
+
+    /**
+     * Creates a String representation of the the formatted number range.
+     *
+     * @return a String containing the localized number range.
+     * @draft ICU 63
+     * @provisional This API might change or be removed in a future release.
+     * @see NumberRangeFormatter
+     */
+    @Override
+    public String toString() {
+        return string.toString();
+    }
+
+    /**
+     * Append the formatted number range to an Appendable, such as a StringBuilder. This may be slightly more efficient
+     * than creating a String.
+     *
+     * <p>
+     * If an IOException occurs when appending to the Appendable, an unchecked {@link ICUUncheckedIOException} is thrown
+     * instead.
+     *
+     * @param appendable
+     *            The Appendable to which to append the formatted number range string.
+     * @return The same Appendable, for chaining.
+     * @draft ICU 63
+     * @provisional This API might change or be removed in a future release.
+     * @see Appendable
+     * @see NumberRangeFormatter
+     */
+    public <A extends Appendable> A appendTo(A appendable) {
```

Comment:
I suggest elevating this into a general utility, since we could use it all over the place, not just with numbers.

Reply:
Hmm.  There is no base class for this.  I suppose we could make a pure-virtual class (interface in Java) called `FormattedValue` with the four main functions:

- toString()
- appendTo()
- nextFieldPosition()
- toCharacterIterator()

Then both this class (FormattedNumberRange) and FormattedNumber would want to implement that interface.

It saves on APIDocs and cleans things up a bit.  It would also be an API change, though, and I'm not sure if I see a strong use case for this particular class being abstracted.  Since this is the second implementation, this would be the time to add such a parent class.  I'll send an email; it's worth asking.

Reply:
I realized that there is more design work required for this.  I filed a ticket:

https://unicode-org.atlassian.net/browse/ICU-20138

I think we should leave this as-is for now because adding a new parent class shouldn't break source compatibility.